guides/plugins/plugins/framework/data-handling/add-data-indexer.md

```diff
@@ -0,0 +1,263 @@
+    * The `update()` can also be used to update data that has always to be changed synchronously.
+* `public function handle(EntityIndexingMessage $message): void`
+    * The `handle()` method handles the messages which were generated in the `self::iterate` or `self::update` function. In the example above a small log entry is written to the database indicating that a customer was indexed. The preferred way to manipulate data here is using the `connection` directly and not to use the DAL. See the section [Use DAL functionalities in the indexer](Use DAL functionalities in the indexer) for more information. 
+    
+### Handle messages asynchronously or synchronously
+By default, all messages which are returned by the `public function update()` function in the indexer are handled synchronously. That means the `handle()` function is called directly after the `update()` function. To handle the messages asynchronously over the [message queue](../../../../hosting/infrastructure/message-queue.md) the `EntityIndexingMessage` can be used with different constructor parameters. A closer look at the `EntityIndexingMessage` class shows that it has a forth parameter named `$forceQueue` which is `false` by default. This parameter can be set to `true` and then the message will be handled asynchronously by the message queue.
+
+### Use DAL functionalities in the indexer
+By default, indexing is also active while working with an indexer, which means, that entities that are written over the DAL also trigger `EntityWrittenContainerEvent` events. So the indexers are triggered again. This can lead to an infinite loop. Therefore, the connection should be used directly to alter data in the database. However, if you want to use the DAL for manipulation data in a data indexer, indexing can be disabled. This can be done by passing adding a flag to the context, as shown in the example below:
+```php
+public function update(EntityWrittenContainerEvent $event): ?EntityIndexingMessage
+{
+    $updates = $event->getPrimaryKeys(CustomerDefinition::ENTITY_NAME);
+
+    if (empty($updates)) {
+        return null;
+    }
+
+    $context = $event->getContext();
+    $context->addExtension(EntityIndexerRegistry::DISABLE_INDEXING, new ArrayEntity());
+
+    return new EntityIndexingMessage(array_values($updates), null, $context);
+}
+```
+
+## Index data using existing events
```

Comment:
Link https://github.com/shopware/platform/blob/trunk/adr/2021-05-14-when-to-use-plain-sql-or-dal.md somewhere